File: src/Azdo.Tui/Components/ErrorModal.cs
using System.Text;
using Azdo.Tui.Rendering;
using Azdo.Tui.Runtime;
using StyleSet = Azdo.Tui.Styles.Styles;

namespace Azdo.Tui.Components;

/// <summary>
/// A message any view can emit to signal a critical API error to the root model,
/// which shows the error modal overlay (≈ <c>components.CriticalErrorMsg</c>).
/// </summary>
public sealed record CriticalErrorMsg(string Title, string Message, string Hint) : IMsg;

/// <summary>Classified error information for display in the error modal (≈ <c>components.ErrorInfo</c>).</summary>
public sealed record ErrorInfo(string Title, string Message, string Hint);

/// <summary>
/// Examines an exception and returns an <see cref="ErrorInfo"/> for known critical
/// errors that should be shown in the error modal (≈ <c>components.ClassifyError</c>).
/// Returns <c>null</c> for transient or unknown errors. Matches on the exception
/// message only so it stays independent of Core types.
/// </summary>
public static class ErrorClassifier
{
    public static ErrorInfo? Classify(Exception? err)
    {
        if (err is null) return null;

        string msg = err.Message;

        if (msg.Contains("HTTP 404") || msg.Contains("resource not found"))
            return new ErrorInfo(
                "Configuration Error",
                "The API returned 'not found'. Your organization or project name in the configuration may be incorrect.",
                "Check your config file and verify the organization and project names match your Azure DevOps setup.");

        if (msg.Contains("HTTP 401") || msg.Contains("authentication failed"))
            return new ErrorInfo(
                "Authentication Error",
                "Your Personal Access Token (PAT) may be expired or invalid.",
                "Run 'azdo auth' to update your PAT.");

        // NOTE: a 403 (insufficient permissions) is intentionally NOT classified
        // as a critical, modal-worthy error. It only means the PAT is missing the
        // scope for one feature (e.g. Build Read for pipelines); the features whose
        // scopes ARE present must keep working. Callers surface 403s inline in the
        // affected tab via the fall-through path. See <see cref="IsPermissionError"/>.

        if (msg.Contains("HTTP 400") || msg.Contains("HTTP request failed with status"))
            return new ErrorInfo(
                "Configuration Error",
                "The API returned an error. Your organization or project name in the configuration may be incorrect.",
                "Check your config file and verify the organization and project names match your Azure DevOps setup.");

        return null;
    }

    /// <summary>
    /// True when the error is a permission failure (HTTP 403): the PAT is valid
    /// but lacks the scope required for one specific feature. Such errors are
    /// surfaced inline in the affected tab and must never block the whole app,
    /// so the features whose scopes are present keep working.
    /// </summary>
    public static bool IsPermissionError(Exception? err)
    {
        if (err is null) return false;
        var msg = err.Message;
        return msg.Contains("HTTP 403") || msg.Contains("access denied");
    }

    /// <summary>
    /// Returns a command that emits a <see cref="CriticalErrorMsg"/> when the error
    /// is critical; otherwise <c>null</c> (≈ <c>components.NewCriticalErrorCmd</c>).
    /// </summary>
    public static Cmd? NewCriticalErrorCmd(Exception? err)
    {
        var info = Classify(err);
        if (info is null) return null;
        return Commands.Of(new CriticalErrorMsg(info.Title, info.Message, info.Hint));
    }
}

/// <summary>
/// An overlay that displays critical error messages (≈ <c>components.ErrorModal</c>).
/// Dismiss on "esc" or "q".
/// </summary>
public sealed class ErrorModal
{
    private const int MinErrorModalWidth = 50;
    private const int ModalHorizontalOverhead = 6;

    private readonly StyleSet _styles;
    private bool _visible;
    private int _width;
    private int _height;
    private string _title = "";
    private string _message = "";
    private string _hint = "";

    public ErrorModal(StyleSet styles) => _styles = styles;

    public void Show(string title, string message, string hint)
    {
        _title = title;
        _message = message;
        _hint = hint;
        _visible = true;
    }

    public void Hide() => _visible = false;
    public bool IsVisible => _visible;
    public void SetSize(int width, int height) { _width = width; _height = height; }

    public Cmd? Update(IMsg msg)
    {
        if (!_visible) return null;
        if (msg is KeyMsg key)
        {
            switch (key.Key)
            {
                case "esc":
                case "q":
                    Hide();
                    return null;
            }
        }
        return null;
    }

    public string View()
    {
        if (!_visible) return "";

        int contentWidth = MinErrorModalWidth;
        if (_title.Length > contentWidth) contentWidth = _title.Length;
        if (_message.Length > contentWidth) contentWidth = _message.Length;

        if (_width > 0)
        {
            int maxContentWidth = _width - ModalHorizontalOverhead;
            if (maxContentWidth < 0) maxContentWidth = 0;
            if (contentWidth > maxContentWidth) contentWidth = maxContentWidth;
        }

        var theme = _styles.Theme;

        var modalStyle = Style.New()
            .WithBorder(Border.Rounded)
            .BorderForeground(theme.Error)
            .Padding(1, 2)
            .Background(theme.BackgroundAlt);

        var titleStyle = Style.New()
            .Foreground(theme.Error)
            .Bold()
            .Width(contentWidth)
            .Background(theme.BackgroundAlt);

        var messageStyle = Style.New()
            .Foreground(theme.Foreground)
            .Width(contentWidth)
            .Background(theme.BackgroundAlt);

        var hintStyle = Style.New()
            .Foreground(theme.Accent)
            .Bold()
            .Width(contentWidth)
            .Background(theme.BackgroundAlt);

        var footerStyle = Style.New()
            .Foreground(theme.ForegroundMuted)
            .Width(contentWidth)
            .Background(theme.BackgroundAlt);

        var content = new StringBuilder();

        // titleStyle has MarginBottom(1) in Go -> emulate with an extra blank line.
        content.Append(titleStyle.Render(_title));
        content.Append('\n');
        content.Append('\n');
        content.Append(messageStyle.Render(_message));
        content.Append('\n');

        if (_hint != "")
        {
            // hintStyle has MarginTop(1) in Go -> emulate with a leading blank line.
            content.Append('\n');
            content.Append(hintStyle.Render(_hint));
            content.Append('\n');
        }

        content.Append('\n');
        content.Append(footerStyle.Render("Press esc to dismiss"));

        var modal = modalStyle.Render(content.ToString());

        if (_width > 0 && _height > 0)
            return Layout.Place(_width, _height, HAlign.Center, VAlign.Center, modal);

        return modal;
    }
}

File: reference/internal/ui/components/errormodal.go
package components

import (
	"strings"

	"github.com/Elpulgo/azdo/internal/ui/styles"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// minErrorModalWidth is the minimum width for the error modal content.
const minErrorModalWidth = 50

// modalHorizontalOverhead accounts for the border (1 char each side)
// and padding (2 chars each side) in the error modal.
const modalHorizontalOverhead = 6

// ErrorInfo holds classified error information for display in the error modal.
type ErrorInfo struct {
	Title   string
	Message string
	Hint    string
}

// ErrorModal is an overlay that displays critical error messages.
type ErrorModal struct {
	styles  *styles.Styles
	visible bool
	width   int
	height  int
	title   string
	message string
	hint    string
}

// NewErrorModal creates a new ErrorModal.
func NewErrorModal(s *styles.Styles) *ErrorModal {
	return &ErrorModal{
		styles:  s,
		visible: false,
	}
}

// Show makes the error modal visible with the given content.
func (m *ErrorModal) Show(title, message, hint string) {
	m.title = title
	m.message = message
	m.hint = hint
	m.visible = true
}

// Hide hides the error modal.
func (m *ErrorModal) Hide() {
	m.visible = false
}

// IsVisible returns true if the modal is visible.
func (m *ErrorModal) IsVisible() bool {
	return m.visible
}

// SetSize sets the available size for the modal.
func (m *ErrorModal) SetSize(width, height int) {
	m.width = width
	m.height = height
}

// Update handles key events for the error modal.
func (m *ErrorModal) Update(msg tea.Msg) (*ErrorModal, tea.Cmd) {
	if !m.visible {
		return m, nil
	}

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.String() {
		case "esc", "q":
			m.Hide()
			return m, nil
		}
	}

	return m, nil
}

// View renders the error modal overlay.
func (m *ErrorModal) View() string {
	if !m.visible {
		return ""
	}

	contentWidth := minErrorModalWidth
	if len(m.title) > contentWidth {
		contentWidth = len(m.title)
	}
	if len(m.message) > contentWidth {
		contentWidth = len(m.message)
	}

	// Cap content width to fit within the available screen width
	if m.width > 0 {
		maxContentWidth := m.width - modalHorizontalOverhead
		if maxContentWidth < 0 {
			maxContentWidth = 0
		}
		if contentWidth > maxContentWidth {
			contentWidth = maxContentWidth
		}
	}

	// Build styles from theme using error color for border
	modalStyle := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(lipgloss.Color(string(m.styles.Theme.Error))).
		Padding(1, 2).
		Background(lipgloss.Color(string(m.styles.Theme.BackgroundAlt)))

	titleStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(string(m.styles.Theme.Error))).
		Bold(true).
		MarginBottom(1).
		Width(contentWidth).
		Background(lipgloss.Color(string(m.styles.Theme.BackgroundAlt)))

	messageStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(string(m.styles.Theme.Foreground))).
		Width(contentWidth).
		Background(lipgloss.Color(string(m.styles.Theme.BackgroundAlt)))

	hintStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(string(m.styles.Theme.Accent))).
		Bold(true).
		MarginTop(1).
		Width(contentWidth).
		Background(lipgloss.Color(string(m.styles.Theme.BackgroundAlt)))

	footerStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(string(m.styles.Theme.ForegroundMuted))).
		Width(contentWidth).
		Background(lipgloss.Color(string(m.styles.Theme.BackgroundAlt)))

	var content strings.Builder

	content.WriteString(titleStyle.Render(m.title))
	content.WriteString("\n")
	content.WriteString(messageStyle.Render(m.message))
	content.WriteString("\n")

	if m.hint != "" {
		content.WriteString(hintStyle.Render(m.hint))
		content.WriteString("\n")
	}

	content.WriteString("\n")
	content.WriteString(footerStyle.Render("Press esc to dismiss"))

	modal := modalStyle.Render(content.String())

	// Center the modal on screen
	if m.width > 0 && m.height > 0 {
		modalWidth := lipgloss.Width(modal)
		modalHeight := lipgloss.Height(modal)

		leftPad := (m.width - modalWidth) / 2
		topPad := (m.height - modalHeight) / 2

		if leftPad < 0 {
			leftPad = 0
		}
		if topPad < 0 {
			topPad = 0
		}

		var centered strings.Builder
		for i := 0; i < topPad; i++ {
			centered.WriteString("\n")
		}

		lines := strings.Split(modal, "\n")
		for _, line := range lines {
			centered.WriteString(strings.Repeat(" ", leftPad))
			centered.WriteString(line)
			centered.WriteString("\n")
		}

		return centered.String()
	}

	return modal
}

// CriticalErrorMsg is a tea.Msg that any view can return as a command
// to signal a critical API error to the root app model, which will show
// the error modal overlay.
type CriticalErrorMsg struct {
	Title   string
	Message string
	Hint    string
}

// NewCriticalErrorCmd checks if err is a critical error and, if so, returns
// a tea.Cmd that emits a CriticalErrorMsg. Returns nil if the error is not critical.
func NewCriticalErrorCmd(err error) tea.Cmd {
	info := ClassifyError(err)
	if info == nil {
		return nil
	}
	return func() tea.Msg {
		return CriticalErrorMsg{
			Title:   info.Title,
			Message: info.Message,
			Hint:    info.Hint,
		}
	}
}

// ClassifyError examines an error and returns an ErrorInfo if it represents
// a critical/known error that should be shown in the error modal.
// Returns nil for transient or unknown errors.
func ClassifyError(err error) *ErrorInfo {
	if err == nil {
		return nil
	}

	msg := err.Error()

	if strings.Contains(msg, "HTTP 404") || strings.Contains(msg, "resource not found") {
		return &ErrorInfo{
			Title:   "Configuration Error",
			Message: "The API returned 'not found'. Your organization or project name in the configuration may be incorrect.",
			Hint:    "Check your config file and verify the organization and project names match your Azure DevOps setup.",
		}
	}

	if strings.Contains(msg, "HTTP 401") || strings.Contains(msg, "authentication failed") {
		return &ErrorInfo{
			Title:   "Authentication Error",
			Message: "Your Personal Access Token (PAT) may be expired or invalid.",
			Hint:    "Run 'azdo auth' to update your PAT.",
		}
	}

	if strings.Contains(msg, "HTTP 403") || strings.Contains(msg, "access denied") {
		return &ErrorInfo{
			Title:   "Authentication Error",
			Message: "Your PAT does not have sufficient permissions for this operation.",
			Hint:    "Run 'azdo auth' to update your PAT with the required scopes.",
		}
	}

	if strings.Contains(msg, "HTTP 400") || strings.Contains(msg, "HTTP request failed with status") {
		return &ErrorInfo{
			Title:   "Configuration Error",
			Message: "The API returned an error. Your organization or project name in the configuration may be incorrect.",
			Hint:    "Check your config file and verify the organization and project names match your Azure DevOps setup.",
		}
	}

	return nil
}
